In [2]:
import pandas as pd
import numpy as np
import country_converter as cococ
from pathlib import Path

In [3]:
DATA = Path(r'C:\Users\indra\Desktop\Thesis\Final project for thesis\Data')

consumption    = DATA / 'National_Fossil_Carbon_Emissions_2025_consumption_emission.xlsx'
population     = DATA / 'Population_with_ISO3.csv'
gdp_per_capita = DATA / 'API_NY.GDP.PCAP.PP.KD_DS2_en_csv_v2_33608.csv'
trade_openess  = DATA / 'API_NE.TRD.GNFS.ZS_DS2_en_csv_v2_171.csv'

out_final = DATA / 'data_final.csv'


EDGAR = DATA / 'EDGAR_1970_2024_production_emission.xlsx'

In [4]:
pd.set_option('display.max_rows', None)

## Consumption and production data cleaning, converting to long format, adding iso3 code and integration

In [5]:
df_c= pd.read_excel(consumption, sheet_name='Consumption Emissions', skiprows=8)
df_p= pd.read_excel(consumption, sheet_name='Territorial Emissions', skiprows=11)

In [6]:
df_c.head()

,Unnamed: 0,Afghanistan,Albania,Algeria,Andorra,Angola,Anguilla,Antigua and Barbuda,Argentina,Armenia,...,Central America,Europe,Middle East,North America,Oceania,South America,International Shipping,International Aviation,Statistical Difference,World
0,1990,NaN,1.521142,NaN,NaN,NaN,NaN,NaN,31.534647,1.718126,...,27.651485,2160.772537,236.166776,1603.565787,75.966392,161.764237,97.945741,51.581993,1.534772e-12,6204.187685
1,1991,NaN,1.259848,NaN,NaN,NaN,NaN,NaN,33.842289,1.323240,...,26.479337,2116.915514,256.960227,1587.351971,76.267238,166.615843,94.244590,53.863925,1.304380e+02,6333.403645
2,1992,NaN,0.798289,NaN,NaN,NaN,NaN,NaN,35.863287,1.039043,...,25.889646,1834.040991,288.842443,1652.287734,77.435693,172.735292,91.447095,59.703935,-1.989520e-13,6146.403948
3,1993,NaN,0.730026,NaN,NaN,NaN,NaN,NaN,34.552463,0.516293,...,26.693117,1813.570118,297.916326,1660.811385,77.655632,179.642928,90.458482,61.679735,0.000000e+00,6209.026223
4,1994,NaN,0.645564,NaN,NaN,NaN,NaN,NaN,34.683976,0.514092,...,27.808745,1769.092533,306.828160,1691.698688,80.418479,180.865685,94.105042,65.509969,5.115908e-13,6270.359809


In [ ]:
drop_regions = [
    'World', 'Statistical Difference', 'International Shipping', 'International Aviation',
    'Africa', 'Asia', 'Europe', 'North America', 'South America', 'Oceania',
    'Central America', 'Middle East', 'OECD', 'Non-OECD', 'EU27',
    'KP Annex B', 'Non KP Annex B',  
]

In [8]:
df_c= df_c.drop(columns=drop_regions)
df_p= df_p.drop(columns=drop_regions)


In [9]:
df_c=df_c.rename(columns={'Unnamed: 0': 'Year'})
df_p=df_p.rename(columns={'Unnamed: 0': 'Year'})
#df_c=df_c.dropna()
#df_p=df_p.dropna()

In [10]:
df_c.head()

,Year,Afghanistan,Albania,Algeria,Andorra,Angola,Anguilla,Antigua and Barbuda,Argentina,Armenia,...,USA,Uruguay,Uzbekistan,Vanuatu,Venezuela,Viet Nam,Wallis and Futuna Islands,Yemen,Zambia,Zimbabwe
0,1990,NaN,1.521142,NaN,NaN,NaN,NaN,NaN,31.534647,1.718126,...,1378.640828,1.311983,NaN,NaN,23.525387,6.353081,NaN,NaN,0.431041,3.513608
1,1991,NaN,1.259848,NaN,NaN,NaN,NaN,NaN,33.842289,1.323240,...,1357.636447,1.466874,NaN,NaN,24.473680,6.322552,NaN,NaN,0.505207,3.470919
2,1992,NaN,0.798289,NaN,NaN,NaN,NaN,NaN,35.863287,1.039043,...,1415.812776,1.660174,NaN,NaN,23.695246,6.320101,NaN,NaN,0.547335,3.661460
3,1993,NaN,0.730026,NaN,NaN,NaN,NaN,NaN,34.552463,0.516293,...,1426.606297,1.526427,NaN,NaN,27.428083,6.867454,NaN,NaN,0.569660,3.316654
4,1994,NaN,0.645564,NaN,NaN,NaN,NaN,NaN,34.683976,0.514092,...,1452.640773,1.463222,NaN,NaN,25.361186,8.035245,NaN,NaN,0.571513,3.454661


**Checking the Consumption Emission** `Year` is now a renamed.

In [11]:
df_p.head()

,Year,Afghanistan,Albania,Algeria,Andorra,Angola,Anguilla,Antigua and Barbuda,Argentina,Armenia,...,USA,Uruguay,Uzbekistan,Vanuatu,Venezuela,Viet Nam,Wallis and Futuna Islands,Yemen,Zambia,Zimbabwe
0,1850,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,0.000069,...,5.402,NaN,0.000807,NaN,NaN,0.0,NaN,NaN,NaN,NaN
1,1851,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,0.000000,...,6.723,NaN,0.000000,NaN,NaN,0.0,NaN,NaN,NaN,NaN
2,1852,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,0.000000,...,7.312,NaN,0.000000,NaN,NaN,0.0,NaN,NaN,NaN,NaN
3,1853,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,0.000000,...,8.232,NaN,0.000000,NaN,NaN,0.0,NaN,NaN,NaN,NaN
4,1854,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,0.000000,...,9.050,NaN,0.000000,NaN,NaN,0.0,NaN,NaN,NaN,NaN


**Checking the Production Emission.** Same number of columns as the consumption
emission, but the series starts in 1850. 

In [ ]:

C_TO_CO2 = 3.664


def to_long_format(df, value_name):
    
    return df.melt(id_vars='Year', var_name='Country', value_name=value_name)


def carbon_to_co2(df, value_name):
    
    df = df.copy()
    df[value_name] = df[value_name] * C_TO_CO2
    return df


df_c_long = carbon_to_co2(to_long_format(df_c, 'Consumption_Emissions'), 'Consumption_Emissions')
df_p_long = carbon_to_co2(to_long_format(df_p, 'Production_Emissions'), 'Production_Emissions')


**On the factor:** GCP reports Mt of *carbon*; this thesis reports Mt of *CO₂*.
The stoichiometric ratio is 44/12 = 3.6667, but GCB publishes 3.664, so 3.664 is
kept to stay consistent with the source. Every emissions figure downstream —
`transfer` included — is therefore in **Mt CO₂**.

In [13]:
df_c_long.head()

,Year,Country,Consumption_Emissions
0,1990,Afghanistan,NaN
1,1991,Afghanistan,NaN
2,1992,Afghanistan,NaN
3,1993,Afghanistan,NaN
4,1994,Afghanistan,NaN


**Consumption in long format.** One row per country-year. The leading `NaN`s are
expected: melting a wide frame materialises every country × year combination,
including the ones with no data.

In [14]:
df_p_long.head()

,Year,Country,Production_Emissions
0,1850,Afghanistan,NaN
1,1851,Afghanistan,NaN
2,1852,Afghanistan,NaN
3,1853,Afghanistan,NaN
4,1854,Afghanistan,NaN


**Production in long format.** Same structure, far more rows, because this series
reaches back to 1850.

In [15]:
print("Shape of df_c_long:", df_c_long.shape)
print("Shape of df_p_long:", df_p_long.shape)
print("Data types of df_c_long:", df_c_long.dtypes)
print("Data types of df_p_long:", df_p_long.dtypes)

Shape of df_c_long: (7490, 3)
Shape of df_p_long: (37450, 3)
Data types of df_c_long: Year                       int64
Country                      str
Consumption_Emissions    float64
dtype: object
Data types of df_p_long: Year                      int64
Country                     str
Production_Emissions    float64
dtype: object


**Pre-cleaning size check.** 7,560 vs 37,800 rows — the gap is almost entirely
the extra 140 years of production history plus the empty combinations from the melt.

In [17]:
print("Shape of df_c_long:", df_c_long.shape)
print("Shape of df_p_long:", df_p_long.shape)
print("Data types of df_c_long:", df_c_long.dtypes)
print("Data types of df_p_long:", df_p_long.dtypes)


Shape of df_c_long: (7490, 3)
Shape of df_p_long: (37450, 3)
Data types of df_c_long: Year                       int64
Country                      str
Consumption_Emissions    float64
dtype: object
Data types of df_p_long: Year                      int64
Country                     str
Production_Emissions    float64
dtype: object


**Post-cleaning size check.** 7,560 → 4,135 and 37,800 → 25,385.

In [18]:
df_c_long.head()

,Year,Country,Consumption_Emissions
0,1990,Afghanistan,NaN
1,1991,Afghanistan,NaN
2,1992,Afghanistan,NaN
3,1993,Afghanistan,NaN
4,1994,Afghanistan,NaN


**Spot-check.** Albania 1990 survives with a real value, confirming the drop hit
only the empty rows.

In [19]:
cc = cococ.CountryConverter()

for df in (df_c_long, df_p_long):
    df['ISO3'] = cc.convert(names=df['Country'], to='ISO3')

unmatched_c = df_c_long.loc[df_c_long['ISO3'].isna(), 'Country'].unique()
print('Unmatched in consumption:', unmatched_c)

unmatched_p = df_p_long.loc[df_p_long['ISO3'].isna(), 'Country'].unique()
print('Unmatched in production:', unmatched_p)

Unmatched in consumption: <StringArray>
[]
Length: 0, dtype: str
Unmatched in production: <StringArray>
[]
Length: 0, dtype: str


In [21]:
print(df_c_long.loc[df_c_long['Consumption_Emissions'].isna(), ['Country', 'Year']])


                                Country  Year
0                           Afghanistan  1990
1                           Afghanistan  1991
2                           Afghanistan  1992
3                           Afghanistan  1993
4                           Afghanistan  1994
5                           Afghanistan  1995
6                           Afghanistan  1996
7                           Afghanistan  1997
8                           Afghanistan  1998
9                           Afghanistan  1999
10                          Afghanistan  2000
11                          Afghanistan  2001
12                          Afghanistan  2002
13                          Afghanistan  2003
14                          Afghanistan  2004
15                          Afghanistan  2005
16                          Afghanistan  2006
17                          Afghanistan  2007
18                          Afghanistan  2008
19                          Afghanistan  2009
20                          Afghan

**Attach ISO3 codes.** The four sources spell countries differently
("Viet Nam" vs "Vietnam", "Korea, Rep." vs "South Korea"), so name-based merging
would silently lose countries. ISO3 is the join key everywhere downstream.


In [22]:
df_c_long.dropna(subset=['Consumption_Emissions'], inplace=True)
df_p_long.dropna(subset=['Production_Emissions'], inplace=True)

In [23]:
df_c_long.head()

,Year,Country,Consumption_Emissions,ISO3
35,1990,Albania,5.573464,ALB
36,1991,Albania,4.616082,ALB
37,1992,Albania,2.924931,ALB
38,1993,Albania,2.674815,ALB
39,1994,Albania,2.365348,ALB


**Consumption data with ISO3.** Albania → `ALB`.

In [24]:
df_p_long.head()

,Year,Country,Production_Emissions,ISO3
99,1949,Afghanistan,0.014656,AFG
100,1950,Afghanistan,0.084272,AFG
101,1951,Afghanistan,0.091600,AFG
102,1952,Afghanistan,0.091600,AFG
103,1953,Afghanistan,0.106256,AFG


In [25]:
df_c_long = df_c_long[df_c_long['Year'].between(1990, 2024)]
df_p_long = df_p_long[df_p_long['Year'].between(1990, 2024)]

**Restrict to 1990–2024.** 

In [26]:
df_p_long.head(5)

,Year,Country,Production_Emissions,ISO3
140,1990,Afghanistan,2.024326,AFG
141,1991,Afghanistan,1.914301,AFG
142,1992,Afghanistan,1.482054,AFG
143,1993,Afghanistan,1.486943,AFG
144,1994,Afghanistan,1.453829,AFG


In [27]:
df_c_long.head(5)

,Year,Country,Consumption_Emissions,ISO3
35,1990,Albania,5.573464,ALB
36,1991,Albania,4.616082,ALB
37,1992,Albania,2.924931,ALB
38,1993,Albania,2.674815,ALB
39,1994,Albania,2.365348,ALB


In [28]:
df=df_c_long.merge(df_p_long, on=['Year', 'ISO3'], how='inner')

**Join the two emission series.** An inner join on `Year`, `ISO3` *and* `Country`
keeps only country-years present in both. That is deliberate: `transfer` is a
difference between the two measures, so a row missing either side is undefined.

The inner join is also what makes the consumption data the binding constraint —
the resulting panel is roughly the size of `df_c_long`.

In [29]:
df.info()   

<class 'pandas.DataFrame'>
RangeIndex: 4066 entries, 0 to 4065
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Year                   4066 non-null   int64  
 1   Country_x              4066 non-null   str    
 2   Consumption_Emissions  4066 non-null   float64
 3   ISO3                   4066 non-null   str    
 4   Country_y              4066 non-null   str    
 5   Production_Emissions   4066 non-null   float64
dtypes: float64(2), int64(1), str(3)
memory usage: 190.7 KB


**Merged panel.** 4,134 country-year rows, no nulls in either emissions column.

In [30]:
df.head(5)

,Year,Country_x,Consumption_Emissions,ISO3,Country_y,Production_Emissions
0,1990,Albania,5.573464,ALB,Albania,5.520602
1,1991,Albania,4.616082,ALB,Albania,4.290226
2,1992,Albania,2.924931,ALB,Albania,2.517116
3,1993,Albania,2.674815,ALB,Albania,2.337999
4,1994,Albania,2.365348,ALB,Albania,1.923806


In [31]:
df.head()

,Year,Country_x,Consumption_Emissions,ISO3,Country_y,Production_Emissions
0,1990,Albania,5.573464,ALB,Albania,5.520602
1,1991,Albania,4.616082,ALB,Albania,4.290226
2,1992,Albania,2.924931,ALB,Albania,2.517116
3,1993,Albania,2.674815,ALB,Albania,2.337999
4,1994,Albania,2.365348,ALB,Albania,1.923806


In [32]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4066 entries, 0 to 4065
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Year                   4066 non-null   int64  
 1   Country_x              4066 non-null   str    
 2   Consumption_Emissions  4066 non-null   float64
 3   ISO3                   4066 non-null   str    
 4   Country_y              4066 non-null   str    
 5   Production_Emissions   4066 non-null   float64
dtypes: float64(2), int64(1), str(3)
memory usage: 190.7 KB


## GDP/capita data cleaning and conversion to long format

In [33]:
df_gdp_pc= pd.read_csv(gdp_per_capita, skiprows=4)
df_gdp_pc.head()

,Country Name,Country Code,Indicator Name,Indicator Code,1960,1961,1962,1963,1964,1965,...,2017,2018,2019,2020,2021,2022,2023,2024,2025,Unnamed: 70
0,Aruba,ABW,"GDP per capita, PPP (constant 2021 internation...",NY.GDP.PCAP.PP.KD,NaN,NaN,NaN,NaN,NaN,NaN,...,38213.289239,39378.228411,39754.560072,30408.827090,35207.170909,39083.461640,42783.248907,45428.002728,NaN,NaN
1,Africa Eastern and Southern,AFE,"GDP per capita, PPP (constant 2021 internation...",NY.GDP.PCAP.PP.KD,NaN,NaN,NaN,NaN,NaN,NaN,...,4223.188050,4229.839634,4200.531044,3978.133502,4049.969513,4098.126192,4076.224898,4099.889812,4164.679177,NaN
2,Afghanistan,AFG,"GDP per capita, PPP (constant 2021 internation...",NY.GDP.PCAP.PP.KD,NaN,NaN,NaN,NaN,NaN,NaN,...,2952.998916,2902.392113,2927.245144,2769.685745,2144.166570,1981.710168,1983.812620,1964.452289,NaN,NaN
3,Africa Western and Central,AFW,"GDP per capita, PPP (constant 2021 internation...",NY.GDP.PCAP.PP.KD,NaN,NaN,NaN,NaN,NaN,NaN,...,6009.516114,6026.527572,6070.284769,5698.766702,5702.134886,5818.483224,5886.796636,6009.084662,6139.587294,NaN
4,Angola,AGO,"GDP per capita, PPP (constant 2021 internation...",NY.GDP.PCAP.PP.KD,NaN,NaN,NaN,NaN,NaN,NaN,...,10565.167648,10154.838923,9712.979449,8904.949006,8720.187687,8750.993630,8597.538788,8752.797471,8760.190820,NaN


**Load GDP per capita.** World Bank CSVs carry four metadata rows before the
header, hence `skiprows=4`. The indicator is PPP-adjusted and in constant 2021
international dollars, so it is comparable both across countries and over time.

In [34]:
years_to_drop = [str(y) for y in range(1960, 1990)]
df_gdp_pc = df_gdp_pc.drop(columns=['Indicator Name', 'Indicator Code'] + years_to_drop)
df_gdp_pc.head()

,Country Name,Country Code,1990,1991,1992,1993,1994,1995,1996,1997,...,2017,2018,2019,2020,2021,2022,2023,2024,2025,Unnamed: 70
0,Aruba,ABW,33146.670548,34079.230463,34458.142298,34627.743475,35580.533978,35476.409987,34506.533080,35534.274503,...,38213.289239,39378.228411,39754.560072,30408.827090,35207.170909,39083.461640,42783.248907,45428.002728,NaN,NaN
1,Africa Eastern and Southern,AFE,3520.019233,3425.136430,3252.316888,3127.877714,3099.430165,3155.044828,3247.125124,3291.616658,...,4223.188050,4229.839634,4200.531044,3978.133502,4049.969513,4098.126192,4076.224898,4099.889812,4164.679177,NaN
2,Afghanistan,AFG,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2952.998916,2902.392113,2927.245144,2769.685745,2144.166570,1981.710168,1983.812620,1964.452289,NaN,NaN
3,Africa Western and Central,AFW,4165.726363,4101.378889,4090.019417,3928.312331,3812.275413,3774.104495,3841.150018,3894.842611,...,6009.516114,6026.527572,6070.284769,5698.766702,5702.134886,5818.483224,5886.796636,6009.084662,6139.587294,NaN
4,Angola,AGO,8366.412716,8170.249394,7445.438367,5481.763868,5377.990350,5981.532576,6565.864794,6808.336281,...,10565.167648,10154.838923,9712.979449,8904.949006,8720.187687,8750.993630,8597.538788,8752.797471,8760.190820,NaN


**Trim to the analysis window.** `Indicator Name`/`Indicator Code` are constant
down the whole file and carry no information; 1960–1989 falls outside the window
set by the emissions data.

In [35]:
df_gdp_pc = df_gdp_pc.drop(columns=['Unnamed: 70'], errors='ignore')

df_gdp_pc_long = df_gdp_pc.melt(
    id_vars=['Country Name', 'Country Code'], var_name='Year', value_name='GDP_per_capita'
)
df_gdp_pc_long['Year'] = df_gdp_pc_long['Year'].astype(int)
df_gdp_pc_long = df_gdp_pc_long.rename(columns={'Country Name': 'Country', 'Country Code': 'ISO3'})
df_gdp_pc_long.head()


,Country,ISO3,Year,GDP_per_capita
0,Aruba,ABW,1990,33146.670548
1,Africa Eastern and Southern,AFE,1990,3520.019233
2,Afghanistan,AFG,1990,NaN
3,Africa Western and Central,AFW,1990,4165.726363
4,Angola,AGO,1990,8366.412716


In [36]:
df_gdp_pc_long = df_gdp_pc_long.sort_values(['Country', 'Year']).reset_index(drop=True)
df_gdp_pc_long.head()

,Country,ISO3,Year,GDP_per_capita
0,Afghanistan,AFG,1990,NaN
1,Afghanistan,AFG,1991,NaN
2,Afghanistan,AFG,1992,NaN
3,Afghanistan,AFG,1993,NaN
4,Afghanistan,AFG,1994,NaN


In [37]:
df_gdp_pc_long.info()

<class 'pandas.DataFrame'>
RangeIndex: 9540 entries, 0 to 9539
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         9540 non-null   str    
 1   ISO3            9540 non-null   str    
 2   Year            9540 non-null   int64  
 3   GDP_per_capita  8669 non-null   float64
dtypes: float64(1), int64(1), str(2)
memory usage: 298.3 KB


**Coverage check.** 9,540 rows, 8,669 non-null. The ~9% gap is countries with
incomplete GDP series and will propagate into the final panel as missing covariates.

## Data Cleaning of trade freeness data set.

In [38]:
df_trade_openess= pd.read_csv(trade_openess, skiprows=4)

In [39]:
df_trade_openess.head()

,Country Name,Country Code,Indicator Name,Indicator Code,1960,1961,1962,1963,1964,1965,...,2017,2018,2019,2020,2021,2022,2023,2024,2025,Unnamed: 70
0,Aruba,ABW,Trade (% of GDP),NE.TRD.GNFS.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,138.716620,138.824972,137.923249,123.240504,139.510903,160.558141,153.719158,164.138645,NaN,NaN
1,Africa Eastern and Southern,AFE,Trade (% of GDP),NE.TRD.GNFS.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,51.760695,54.868090,53.214310,49.343585,54.551332,57.117376,56.283721,54.203271,51.994817,NaN
2,Afghanistan,AFG,Trade (% of GDP),NE.TRD.GNFS.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,46.709895,51.411716,72.885470,66.999578,83.763025,NaN,NaN
3,Africa Western and Central,AFW,Trade (% of GDP),NE.TRD.GNFS.ZS,33.02326,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Angola,AGO,Trade (% of GDP),NE.TRD.GNFS.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,44.996325,58.269895,60.371307,54.660403,62.528771,56.538072,53.607013,49.848951,34.646640,NaN


In [40]:
df_trade_openess = df_trade_openess.drop(columns=['Indicator Name', 'Indicator Code'] + years_to_drop)
df_trade_openess.head()

,Country Name,Country Code,1990,1991,1992,1993,1994,1995,1996,1997,...,2017,2018,2019,2020,2021,2022,2023,2024,2025,Unnamed: 70
0,Aruba,ABW,NaN,NaN,NaN,NaN,NaN,171.362098,175.344130,168.781911,...,138.716620,138.824972,137.923249,123.240504,139.510903,160.558141,153.719158,164.138645,NaN,NaN
1,Africa Eastern and Southern,AFE,41.287094,39.781486,40.460789,42.900951,45.509861,47.491919,47.254615,45.115559,...,51.760695,54.868090,53.214310,49.343585,54.551332,57.117376,56.283721,54.203271,51.994817,NaN
2,Afghanistan,AFG,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,46.709895,51.411716,72.885470,66.999578,83.763025,NaN,NaN
3,Africa Western and Central,AFW,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Angola,AGO,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,44.996325,58.269895,60.371307,54.660403,62.528771,56.538072,53.607013,49.848951,34.646640,NaN


**Trim.** Reuses `years_to_drop` defined in the GDP section, so both World Bank
files are cut to the same window.

In [41]:
df_trade_openess = df_trade_openess.drop(columns=['Unnamed: 70'], errors='ignore')

df_trade_openess_long = df_trade_openess.melt(
    id_vars=['Country Name', 'Country Code'], var_name='Year', value_name='Trade_Openness'
)
df_trade_openess_long['Year'] = df_trade_openess_long['Year'].astype(int)
df_trade_openess_long = df_trade_openess_long.rename(columns={'Country Name': 'Country', 'Country Code': 'ISO3'})
df_trade_openess_long.head()


,Country,ISO3,Year,Trade_Openness
0,Aruba,ABW,1990,NaN
1,Africa Eastern and Southern,AFE,1990,41.287094
2,Afghanistan,AFG,1990,NaN
3,Africa Western and Central,AFW,1990,NaN
4,Angola,AGO,1990,NaN


**Reshape to long.** Identical treatment to the GDP file, producing the same
`Country`/`ISO3`/`Year` key.

In [42]:
df_trade_openess_long = df_trade_openess_long.sort_values(['Country', 'Year']).reset_index(drop=True)
df_trade_openess_long.head()

,Country,ISO3,Year,Trade_Openness
0,Afghanistan,AFG,1990,NaN
1,Afghanistan,AFG,1991,NaN
2,Afghanistan,AFG,1992,NaN
3,Afghanistan,AFG,1993,NaN
4,Afghanistan,AFG,1994,NaN


In [43]:
df_trade_openess_long.info()

<class 'pandas.DataFrame'>
RangeIndex: 9540 entries, 0 to 9539
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         9540 non-null   str    
 1   ISO3            9540 non-null   str    
 2   Year            9540 non-null   int64  
 3   Trade_Openness  7564 non-null   float64
dtypes: float64(1), int64(1), str(2)
memory usage: 298.3 KB


## Population dataset

In [44]:
df_population= pd.read_csv(population)

In [45]:
df_population = df_population[df_population['Year'] >= 1990]
df_population['Population'] = pd.to_numeric(df_population['Population'], errors='coerce')

In [46]:
df_population.info()

<class 'pandas.DataFrame'>
Index: 7700 entries, 20 to 12099
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Country/Area  7700 non-null   str    
 1   Year          7700 non-null   int64  
 2   Unit          7700 non-null   str    
 3   Population    7407 non-null   float64
 4   ISO3          7490 non-null   str    
dtypes: float64(1), int64(1), str(3)
memory usage: 360.9 KB


In [47]:
parts = (df_population.groupby(['ISO3', 'Year'])['Country/Area'].nunique())
p= parts[parts > 1]
print("Number of unique countries per ISO3 and Year > 1:\n", p)

Number of unique countries per ISO3 and Year > 1:
 ISO3  Year
TZA   1990    2
      1991    2
      1992    2
      1993    2
      1994    2
      1995    2
      1996    2
      1997    2
      1998    2
      1999    2
      2000    2
      2001    2
      2002    2
      2003    2
      2004    2
      2005    2
      2006    2
      2007    2
      2008    2
      2009    2
      2010    2
      2011    2
      2012    2
      2013    2
      2014    2
      2015    2
      2016    2
      2017    2
      2018    2
      2019    2
      2020    2
      2021    2
      2022    2
      2023    2
      2024    2
YEM   1990    3
      1991    3
      1992    3
      1993    3
      1994    3
      1995    3
      1996    3
      1997    3
      1998    3
      1999    3
      2000    3
      2001    3
      2002    3
      2003    3
      2004    3
      2005    3
      2006    3
      2007    3
      2008    3
      2009    3
      2010    3
      2011    3
      2012    3
      2013

## Tanzania and Yemen has duplicated ISO3 and will be checked individually

In [48]:
df_population[df_population['ISO3'] == 'YEM']

,Country/Area,Year,Unit,Population,ISO3
11845,Yemen,1990,Persons,13887862.0,YEM
11846,Yemen,1991,Persons,14430201.0,YEM
11847,Yemen,1992,Persons,14989494.0,YEM
11848,Yemen,1993,Persons,15563914.0,YEM
11849,Yemen,1994,Persons,16149147.0,YEM
11850,Yemen,1995,Persons,16740058.0,YEM
11851,Yemen,1996,Persons,17313177.0,YEM
11852,Yemen,1997,Persons,17873936.0,YEM
11853,Yemen,1998,Persons,18446013.0,YEM
11854,Yemen,1999,Persons,19029420.0,YEM


In [49]:
for iso in p.index.get_level_values('ISO3').unique():
    names = sorted(df_population.loc[df_population['ISO3'] == iso, 'Country/Area'].unique())
    print(f'  {iso}: {names}')

  TZA: ['United Republic of Tanzania: Mainland', 'United Republic of Tanzania: Zanzibar']
  YEM: ['Yemen', 'Yemen: Former Democratic Yemen', 'Yemen: Former Yemen Arab Republic']


In [50]:
tan= df_population[df_population['Country/Area'].isin(['United Republic of Tanzania: Mainland', 'United Republic of Tanzania: Zanzibar'])]

aggregated_tan = tan.groupby(['Year'], as_index=False).agg({'Population': 'sum'})

df_population = pd.concat([df_population[~df_population['ISO3'].isin(['TZA'])], aggregated_tan.assign(ISO3='TZA', **{'Country/Area': 'United Republic of Tanzania'})], ignore_index=True)

In [51]:
df_population.info()

<class 'pandas.DataFrame'>
RangeIndex: 7665 entries, 0 to 7664
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Country/Area  7665 non-null   str    
 1   Year          7665 non-null   int64  
 2   Unit          7630 non-null   str    
 3   Population    7372 non-null   float64
 4   ISO3          7455 non-null   str    
dtypes: float64(1), int64(1), str(3)
memory usage: 299.5 KB


**Structure check.** Note `Population` loads as a **string**, not an integer —
fixed further down, before the data is written out. `ISO3` has 210 nulls, which
are non-sovereign territories and aggregates; the inner joins drop them.

In [52]:
df_population.head(40)

,Country/Area,Year,Unit,Population,ISO3
0,Afghanistan,1990,Persons,12045660.0,AFG
1,Afghanistan,1991,Persons,12238879.0,AFG
2,Afghanistan,1992,Persons,13278974.0,AFG
3,Afghanistan,1993,Persons,14943172.0,AFG
4,Afghanistan,1994,Persons,16250794.0,AFG
5,Afghanistan,1995,Persons,17065836.0,AFG
6,Afghanistan,1996,Persons,17763266.0,AFG
7,Afghanistan,1997,Persons,18452091.0,AFG
8,Afghanistan,1998,Persons,19159996.0,AFG
9,Afghanistan,1999,Persons,19887785.0,AFG


## Merging population gdp/capita and trade freeness together

In [ ]:
df2 = df_gdp_pc_long.merge(df_trade_openess_long, on=['ISO3', 'Year'], how='inner')
df2 = df2.merge(df_population, on=['ISO3', 'Year'], how='inner')

**Assemble the covariates.** GDP per capita, then trade openness, then population —
joined on `ISO3` + `Year` only, never on country name, since the three sources spell
names differently.

These are inner joins, so a country must appear in all three files to survive.
`Country_x`/`Country_y` are the redundant name columns; they are dropped after the
final merge.

The assertion enforces the panel's key invariant: one row per country-year. This is
the check that would have caught the Tanzania duplication at source.

In [54]:
df2 = df2.sort_values(['Country_x', 'Year']).reset_index(drop=True)
df2.head()

,Country_x,ISO3,Year,GDP_per_capita,Country_y,Trade_Openness,Country/Area,Unit,Population
0,Afghanistan,AFG,1990,NaN,Afghanistan,NaN,Afghanistan,Persons,12045660.0
1,Afghanistan,AFG,1991,NaN,Afghanistan,NaN,Afghanistan,Persons,12238879.0
2,Afghanistan,AFG,1992,NaN,Afghanistan,NaN,Afghanistan,Persons,13278974.0
3,Afghanistan,AFG,1993,NaN,Afghanistan,NaN,Afghanistan,Persons,14943172.0
4,Afghanistan,AFG,1994,NaN,Afghanistan,NaN,Afghanistan,Persons,16250794.0


**Sort.** Cosmetic.

## Merging df1 and df2

In [55]:
df1 = df
df_final = df1.merge(df2, on=['ISO3', 'Year'], how='inner')


df_final.head()

,Year,Country_x_x,Consumption_Emissions,ISO3,Country_y_x,Production_Emissions,Country_x_y,GDP_per_capita,Country_y_y,Trade_Openness,Country/Area,Unit,Population
0,1990,Albania,5.573464,ALB,Albania,5.520602,Albania,5560.833935,Albania,39.436963,Albania,Persons,3277966.0
1,1991,Albania,4.616082,ALB,Albania,4.290226,Albania,4027.888807,Albania,36.070520,Albania,Persons,3282602.0
2,1992,Albania,2.924931,ALB,Albania,2.517116,Albania,3761.139808,Albania,108.785472,Albania,Persons,3282501.0
3,1993,Albania,2.674815,ALB,Albania,2.337999,Albania,4145.902505,Albania,80.518333,Albania,Persons,3277825.0
4,1994,Albania,2.365348,ALB,Albania,1.923806,Albania,4517.780079,Albania,53.102585,Albania,Persons,3269410.0


**Final merge.** The emissions panel joined to the covariate panel on `ISO3` + `Year`,
with the same uniqueness guard. Inner again, so the result is the intersection of every
source.

In [56]:
df_final = df_final[[ 'Country/Area', 'ISO3', 'Year', 'Consumption_Emissions', 'Production_Emissions', 'GDP_per_capita', 'Trade_Openness', 'Population']]
df_final = df_final.rename(columns={'Country/Area': 'Country'})
df_final.head()

,Country,ISO3,Year,Consumption_Emissions,Production_Emissions,GDP_per_capita,Trade_Openness,Population
0,Albania,ALB,1990,5.573464,5.520602,5560.833935,39.436963,3277966.0
1,Albania,ALB,1991,4.616082,4.290226,4027.888807,36.070520,3282602.0
2,Albania,ALB,1992,2.924931,2.517116,3761.139808,108.785472,3282501.0
3,Albania,ALB,1993,2.674815,2.337999,4145.902505,80.518333,3277825.0
4,Albania,ALB,1994,2.365348,1.923806,4517.780079,53.102585,3269410.0


**Select and order the analysis columns.** Discards the three duplicate country
name columns and the constant `Unit` column, keeping one clean `Country` label.

In [57]:
df_final.info()


<class 'pandas.DataFrame'>
RangeIndex: 4032 entries, 0 to 4031
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Country                4032 non-null   str    
 1   ISO3                   4032 non-null   str    
 2   Year                   4032 non-null   int64  
 3   Consumption_Emissions  4032 non-null   float64
 4   Production_Emissions   4032 non-null   float64
 5   GDP_per_capita         3998 non-null   float64
 6   Trade_Openness         3805 non-null   float64
 7   Population             4032 non-null   float64
dtypes: float64(5), int64(1), str(2)
memory usage: 252.1 KB


**Pre-cast check.** 4,066 rows. `GDP_per_capita` is 34 short and
`Trade_Openness` 227 short of complete — these gaps are real and need an explicit
decision at estimation time (listwise deletion vs. a restricted sample). Note
`Population` is string which will be converted to int.

In [58]:
df_final['Population'] = pd.to_numeric(df_final['Population'], errors='raise').astype('int64')

In [59]:
df_final.info()

<class 'pandas.DataFrame'>
RangeIndex: 4032 entries, 0 to 4031
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Country                4032 non-null   str    
 1   ISO3                   4032 non-null   str    
 2   Year                   4032 non-null   int64  
 3   Consumption_Emissions  4032 non-null   float64
 4   Production_Emissions   4032 non-null   float64
 5   GDP_per_capita         3998 non-null   float64
 6   Trade_Openness         3805 non-null   float64
 7   Population             4032 non-null   int64  
dtypes: float64(4), int64(2), str(2)
memory usage: 252.1 KB


**Confirm the cast.** `Population` is now `int64`.

# **Treatment Block** 

In [ ]:
ANNEX_B = {
    'AUS','AUT','BEL','BGR','CAN','HRV','CZE','DNK','EST','FIN','FRA','DEU',
    'GRC','HUN','ISL','IRL','ITA','JPN','LVA','LIE','LTU','LUX','MCO','NLD',
    'NZL','NOR','POL','PRT','ROU','RUS','SVK','SVN','ESP','SWE','CHE','UKR','GBR',
}  # USA excluded: as theysigned but never-treated.

# Catch typos/format drift instead of silently coding a treated country as control
missing = ANNEX_B - set(df_final['ISO3'])
assert missing <= {'ISL','LIE','MCO'}, f"unexpected Annex B codes absent: {missing}"

KYOTO_ENTRY_INTO_FORCE  = 2005   # treaty legally binding from 16 Feb 2005
FIRST_COMMITMENT_PERIOD = 2008   # 2008-2012 compliance window

df_final['annex_b']     = df_final['ISO3'].isin(ANNEX_B).astype(int)
df_final['first_treat'] = df_final['annex_b'] * KYOTO_ENTRY_INTO_FORCE  # 0 = never-treated
df_final['post']        = (df_final['Year'] >= KYOTO_ENTRY_INTO_FORCE).astype(int)
df_final['rel_year']    = np.where(df_final['annex_b'] == 1,
                                   df_final['Year'] - KYOTO_ENTRY_INTO_FORCE, np.nan)  # event-study axis

# Same coding anchored on 2008, for the timing robustness check.
df_final['first_treat_2008'] = df_final['annex_b'] * FIRST_COMMITMENT_PERIOD
df_final['post_2008']        = (df_final['Year'] >= FIRST_COMMITMENT_PERIOD).astype(int)
df_final['rel_year_2008']    = np.where(df_final['annex_b'] == 1,
                                        df_final['Year'] - FIRST_COMMITMENT_PERIOD, np.nan)

# C&S needs a numeric unit id
df_final['country_id'] = df_final['ISO3'].map(
    {iso: i for i, iso in enumerate(sorted(df_final['ISO3'].unique()))}
)

print(df_final.groupby('annex_b')['ISO3'].nunique())         
print(df_final.groupby('first_treat')['ISO3'].nunique())     

annex_b
0    85
1    34
Name: ISO3, dtype: int64
first_treat
0       85
2005    34
Name: ISO3, dtype: int64


**Treatment coding for the DiD.**

`ANNEX_B` is the set of countries with binding first-period Kyoto targets. The USA
is deliberately excluded: it signed but never ratified, so it never faced a binding
target and belongs in the never-treated control group. The assertion guards against
a code silently dropping out of the panel — `ISL`, `LIE` and `MCO` are allowed to be
absent because they are too small to appear in all source files.

**On the timing.** Two dates are defensible and the choice is a real modelling
decision, not a detail:

- **2005** — the treaty entered into force on 16 Feb 2005, so this is when the obligation became binding and anticipation effects could start.
- **2008** — the first commitment period (2008–2012) is when compliance was actually measured.

2005 is the main specification; `first_treat_2008`, `post_2008` and `rel_year_2008`
carry the alternative so the robustness check is a one-line change at estimation
rather than a re-run of this notebook.

**A note on the estimator.** `first_treat` follows the Callaway–Sant'Anna
convention (`0` = never-treated). But every treated unit here shares a single
adoption date, so there is no staggered adoption — C&S collapses to essentially a
2×2 design and buys little over a two-way fixed effects model with an event-study
plot on `rel_year`. Both should give the same answer; TWFE is the simpler thing to
defend in the write-up.

In [61]:
df_final.head()

,Country,ISO3,Year,Consumption_Emissions,Production_Emissions,GDP_per_capita,Trade_Openness,Population,annex_b,first_treat,post,rel_year,first_treat_2008,post_2008,rel_year_2008,country_id
0,Albania,ALB,1990,5.573464,5.520602,5560.833935,39.436963,3277966,0,0,0,NaN,0,0,NaN,0
1,Albania,ALB,1991,4.616082,4.290226,4027.888807,36.070520,3282602,0,0,0,NaN,0,0,NaN,0
2,Albania,ALB,1992,2.924931,2.517116,3761.139808,108.785472,3282501,0,0,0,NaN,0,0,NaN,0
3,Albania,ALB,1993,2.674815,2.337999,4145.902505,80.518333,3277825,0,0,0,NaN,0,0,NaN,0
4,Albania,ALB,1994,2.365348,1.923806,4517.780079,53.102585,3269410,0,0,0,NaN,0,0,NaN,0


In [63]:

df_final = df_final.sort_values(['ISO3', 'Year']).reset_index(drop=True)

df_final.to_csv(out_final, index=False)
